File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def solution(s):
    ts = s
    answer = 0
    while True:
        flag = True
        stack = []
        for i in s:
            if not stack:
                if i == ']' or i == ')' or i == '}':
                    flag = False
                    break
                else:
                    stack.append(i)

            else:
                if i == '[' or i == '(' or i == '{':
                    stack.append(i)
                else:
                    if (stack[-1] == '[' and i == ']') or (stack[-1] == '{' and i == '}') or (stack[-1] == '(' and i == ')'):
                        stack.pop()
                    else:
                        flag = False
                        break
        if flag and not stack:
            answer += 1
        
        s = s[1:] + s[0]
        
        if s == ts:
            break
                    
    
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <stack>
using namespace std;

bool check(string s) {
    stack<char> st;
    for (int i = 0; i < s.size(); i++){
        // std::cout<<s<<endl;
        if (st.empty()){
            st.push(s[i]);
        }
        else if (s[i]=='(' || s[i]=='[' || s[i]=='{' ){
            st.push(s[i]);
        }
        else if ((st.top() == '(' && s[i] == ')') || (st.top() == '{' && s[i] == '}') || (st.top() == '[' && s[i] == ']') ){
            st.pop();
        }
        else {
            return false;
        }
    }
    // std::cout<<st.empty()<<endl;
    return st.empty();
}

int solution(string s) {
    int answer = 0;
    for (int i = 0; i < s.size(); i++){
        if (check(s)==1) {
            answer++;
        }
        s = s.substr(1) + s[0];
        }
    if (answer > 0) {
        return answer;
    }
    else return 0;
    }